### PR TITLE
 Fix: Closures in DataStreams prevent pickling

### LINKED ISSUE #157: Closures in DataStreams prevent pickling
## Describe the bug

The DataStream's `from_*` methods all create closures around local functions to provide the `generator_func` argument to the initializer. This prevents `DataStream`s created in this manner from being pickled. Pickling is important for at least one specific use case for us- using `ray` to execute things remotely. Not being able to pickle streams effectively blocks us from being able to write a shared train backend with ray, since most training things involve streams of training data.

## Platform

Please provide details about the environment you are using, including the following:

- Interpreter version: all
- Library version: 0.5

## Sample Code

```
stream = DataStream.from_iterable([1,2,3])
pickle.dumps(stream)
```

## Expected behavior

it should pickle

## Observed behavior

it does not

## Additional context

The `DataStream` class was designed to hold args and kwargs for the generator function as private member variable anyway, so closures aren't required to implement these `from_*` classmethods

### PR DESCRIPTION
**Pull request by kellyaa**
This PR fixes #157 

Authored with copious input from @joerunde

### CODE REVIEW

**Comment by gabe-l-hart** [comment]
@kellyaa To get DCO to pass, you'll need to:

```sh
git rebase origin/main --signoff
git push -f
```

**Review by joerunde — APPROVED** [review]
Looks great!

The only comments I have are more future looking:
- If we're trying to ensure that all of our `from_*` stream constructors are pickleable, should we add tests for every one? On the one hand it seems a bit overkill, on the other we do need this tested somewhere if it's critical to the ray pieces, but maybe we could ensure that's covered when we land the ray training bits.
- There are plenty of other places here where we use lambdas, like `map` and `filter`. Should we address those now as well? Or open an issue about it?
- This is probably worth a small comment in the DataStream class docstring about why we want to avoid lambdas if possible

**Review by gabe-l-hart — CHANGES_REQUESTED** [review]
NITs about proactive log formatting (sorry!) but otherwise it's good to go

File: caikit/core/data_model/streams/data_stream.py [hunk 1]
```diff
@@ -168,28 +170,28 @@ def from_jsonl(cls, filename: str) -> "DataStream[Dict]":
...
-                        "<COR35596551E>",
-                        ValueError("Invalid JSON object in `{}`".format(line)),
-                    )
+        return cls(cls._from_jsonl_generator, filename)
 
-        return cls(generator_func)
+    @classmethod
+    def _from_jsonl_generator(cls, filename):
+        with open(filename, mode="rb") as json_fh:
+            log.debug2("Loading JSON array file: {}".format(filename))
```
[the remark this answers is not in the document]
Comment (kellyaa) [reply caikit/core/data_model/streams/data_stream.py:178]: Will put this in next commit.

File: caikit/core/data_model/streams/data_stream.py [hunk 2]
```diff
@@ -125,9 +125,11 @@ def from_iterable(cls, data: typing.Iterable[T]) -> "DataStream[T]":
             3
         """
         error.type_check("<COR88684982E>", Iterable, data=data)
+        return cls(cls._from_iterable_generator, data)
 
-        # lambda closure over `data` that returns an iterator over `data`
-        return cls(lambda: iter(data))
+    @classmethod
+    def _from_iterable_generator(cls, data: typing.Iterable[T]):
```
[the remark this answers is not in the document]
Comment (kellyaa) [reply caikit/core/data_model/streams/data_stream.py:131]: Will put this in next commit

File: caikit/core/data_model/streams/data_stream.py [hunk 3]
```diff
@@ -337,22 +338,20 @@ def from_header_csv(cls, filename: str, *args, **kwargs) -> "DataStream[Dict]":
             error(
                 "<COR44308234E>",
                 FileNotFoundError(
-                    "csv filename `{}` does not exist or is not a regular file.".format(
-                        filename
-                    )
+                    "csv filename %s does not exist or is not a regular file.", filename
```
Comment (gabe-l-hart) [inline caikit/core/data_model/streams/data_stream.py:341]: Shoot, this actually _doesn't_ work. When instantiating an exception, using `.format` (or the old school `"foo %s" % "bar"` syntax) to explicitly do the formatting is what you want to do. It's only in arguments to logging functions that you want to use `,` to delay the interpolation

**Review by gabe-l-hart — CHANGES_REQUESTED** [review]
I think there are a couple of places where we need to go back to `.format` when explicitly creating error string

**Review by gabe-l-hart — APPROVED** [review]
Looks good!

**Pull request merged**